includes/wikia/services/UserStatsService.class.php

```diff
@@ -485,64 +527,81 @@ public function getFirstContributionTimestamp( $wikiId = 0 ) {
 		wfProfileIn( __METHOD__ );
 
-		$dbName = false;
-		if ( $wikiId != 0 ) {
-			$dbName = WikiFactory::IDtoDB( $wikiId );
-		} else {
-			$wikiId = $this->wg->CityId;
-		}
-
-		$dbr = $this->getWikiDB( DB_SLAVE, $dbName );
-
+		$dbr = $this->getDatabase();
 		$res = $dbr->selectRow(
 			'revision',
-			array('min(rev_timestamp) AS firstContributionTimestamp'),
-			array('rev_user' => $this->userId),
+			[ 'min(rev_timestamp) AS firstContributionTimestamp' ],
+			[ 'rev_user' => $this->userId ],
 			__METHOD__
 		);
+
 		$firstContributionTimestamp = null;
-		if( !empty($res) ) {
+		if( !empty( $res ) ) {
 			$firstContributionTimestamp = $res->firstContributionTimestamp;
-			$this->setOptionWiki( 'firstContributionTimestamp', $firstContributionTimestamp, $wikiId );
+			$this->setOptionWiki( 'firstContributionTimestamp', $firstContributionTimestamp );
 		}
 
 		wfProfileOut( __METHOD__ );
 		return $firstContributionTimestamp;
 	}
 
-
-	/**
-	 * Get timestamp of last (most recent) user's contribution on specified wiki.
-	 * @since Nov 2013
-	 * @author Kamil Koterba
-	 *
-	 * @param $wikiId Integer Id of wiki - specifies wiki from which to get editcount, 0 for current wiki
-	 * @return String Timestamp in format YmdHis e.g. 20131107192200 or null
-	 */
-	private function getLastContributionTimestamp( $wikiId = 0 ) {
-		wfProfileIn( __METHOD__ );
-
-		$dbName = ( $wikiId == 0 ) ? false : WikiFactory::IDtoDB( $wikiId );
-
+	private function initLastContributionTimestamp() {
 		/* Get lastContributionTimestamp from database */
-		$dbr = $this->getWikiDB( DB_SLAVE, $dbName );
-
+		$dbr = $this->getDatabase();
 		$res = $dbr->selectRow(
 			'revision',
-			array('max(rev_timestamp) AS lastContributionTimestamp'),
-			array('rev_user' => $this->userId),
+			[ 'max(rev_timestamp) AS lastContributionTimestamp' ],
+			[ 'rev_user' => $this->userId ],
 			__METHOD__
 		);
 		$lastContributionTimestamp = null;
-		if( !empty($res) ) {
+		if( !empty( $res ) ) {
 			$lastContributionTimestamp = $res->lastContributionTimestamp;
+			$this->setOptionWiki( 'lastContributionTimestamp', $lastContributionTimestamp );
 		}
 
 		wfProfileOut( __METHOD__ );
 		return $lastContributionTimestamp;
 	}
 
+	private function initWikiId( $wikiId ) {
+		$this->wikiId = ( $wikiId === 0 ) ? $this->wg->CityId : $wikiId;
+	}
+
+	private function getWikiId() {
+		return $this->wikiId;
+	}
+
+	private function getDatabase( $flags = 0 ) {
+		$dbName = ( $this->getWikiId() === $this->wg->CityId ) ? false : WikiFactory::IDtoDB( $this->getWikiId() );
+		$dbType = ( $flags & Title::GAID_FOR_UPDATE ) ? DB_MASTER : DB_SLAVE;
+
+		return $this->getWikiDB( $dbType, $dbName );
+	}
+
+	private static function getStatsMemcKey( $userId, $wikiId ) {
+		return wfSharedMemcKey( 'services', 'userstats', 'stats4', $wikiId, $userId );
+	}
+
+	private static function getOptionsWikiMemcKey( $userId, $wikiId ) {
+		return wfMemcKey( 'optionsWiki', $wikiId, $userId );
```

Comment:
I think we could use wfSharedMemcKey here, it's pointless to duplicate cached options on each wiki as you can request these options from other wikis as well.